pennylane/_qubit_device.py

```diff
@@ -112,83 +112,246 @@ def execute(self, queue, observables, parameters=None):
 
             # Ensures that a combination with sample does not put
             # expvals and vars in superfluous arrays
-            if all(obs.return_type is Sample for obs in observables):
-                return self._asarray(results)
-            if any(obs.return_type is Sample for obs in observables):
+            sample_return_types = (obs.return_type is Sample for obs in observables)
+            if any(sample_return_types) and not all(sample_return_types):
                 return self._asarray(results, dtype="object")
 
             return self._asarray(results)
 
-    def rotate_basis(self, observable):
+    def statistics(self, observables):
+        """Process measurement results from circuit execution and return statistics.
+
+        This includes returning expectation values, variance, samples and probabilities.
+
+        Args:
+            observables (Union[:class:`Observable`, List[:class:`Observable`]]): the number of basis states to sample from
+
+        Raises:
+            QuantumFunctionError: if the value of :attr:`~.Observable.return_type` is not supported
+
+        Returns:
+            Union[float, List[float]]: the corresponding statistics
+        """
+        results = []
+
+        for obs in observables:
+            # Pass instances directly
+            if obs.return_type is Expectation:
+                results.append(self.expval(obs))
+
+            elif obs.return_type is Variance:
+                results.append(self.var(obs))
+
+            elif obs.return_type is Sample:
+                results.append(np.array(self.sample(obs)))
+
+            elif obs.return_type is Probability:
+                results.append(self.probability(wires=obs.wires))
+
+            elif obs.return_type is not None:
+                raise QuantumFunctionError("Unsupported return type specified for observable {}".format(obs.name))
+
+        return results
+
+    def rotate_basis(self, obs_queue):
         """Rotates the specified wires such that they
         are in the eigenbasis of the provided observable.
 
         Args:
-            observable (Observable): the observable that is to be
-                measured
+            observables (List[:class:`Observable`]): the observable queue
+        """
+        self._prob = self.probability()
+        self._memory = False
+
+        wires = []
+
+        for observable in obs_queue:
+            if hasattr(observable, "return_type") and observable.return_type == Sample:
+                self._memory = True  # make sure to return samples
+
+            for diag_gate in observable.diagonalizing_gates():
+                self.apply(diag_gate)
+
+            for wire in observable.wires:
+                if isinstance(wire, int):
+                    wires.append(wire)
+                else:
+                    wires.extend(wire)
+
+        self._wires_used = wires
+        self._rotated_prob = self.probability(wires)
```

Comment:
One could think about not calculating the probabilities if there are only samples requested from the circuit

Reply:
As the previous attribute, this would also mostly be something that could be used for sanity check. One samples are being returned, this would be equal to having the probabilities before computational basis samples are being extracted